Update TESQuest, MapMarkerData, PlayerSkills, FileExists

zax-ftw opened the pull request:
Added `ForceRefIntoAliasID`, `GetType` and `GetPriority` to `TESQuest` class. Updated `MapMarkerData` with names taken from decompiled `map.swf` (added more types, updated some typos, more consistency). Made `FileExists` from `BSScaleformManager` public. Added `kNone` to `PlayerSkills` enum for consistency with other enums such as `ActorValues`.

Ryan-rsm-McKenzie commented:
I'm just going to make the maintenance script only run on push since it refuses to work

Ryan-rsm-McKenzie commented on include/RE/B/BSScaleformManager.h:
```diff
@@ -28,6 +28,7 @@ namespace RE
 		using ScaleModeType = GFxMovieView::ScaleModeType;
 
 		static BSScaleformManager* GetSingleton();
+		static bool                FileExists(const char* a_fileName);
```
Why do you need to expose this?

Ryan-rsm-McKenzie replied:
Please order these two alphabetically

Ryan-rsm-McKenzie commented on include/RE/Offsets.h:
```diff
@@ -651,6 +651,7 @@ namespace RE
 		{
 			inline constexpr REL::ID EnsureQuestStarted(static_cast<std::uint64_t>(24481));
 			inline constexpr REL::ID ResetQuest(static_cast<std::uint64_t>(24486));
+			inline constexpr REL::ID ForceRefIntoAliasID(static_cast<std::uint64_t>(24523));
```
`Offsets.h` is deprecated, please don't use it.

zax-ftw replied:
What's the replacement for this file?

Ryan-rsm-McKenzie replied:
There is no replacement, just define it ad-hoc

zax-ftw commented on include/RE/B/BSScaleformManager.h:
```diff
@@ -28,6 +28,7 @@ namespace RE
 		using ScaleModeType = GFxMovieView::ScaleModeType;
 
 		static BSScaleformManager* GetSingleton();
+		static bool                FileExists(const char* a_fileName);
```
To check where other SWFs are located (HUDMenu in my case, it can be either in `/exported` or not).

Ryan-rsm-McKenzie replied:
This function checks the both the root and exported folders, it doesn't do what you want

zax-ftw replied:
They call it for root and exported separately: [https://i.imgur.com/SwQ1hCp.png](url)

Ryan-rsm-McKenzie replied:
The source for these functions are ordered relative to their declarations, please reorder this function in the source file.

zax-ftw replied:
Added requested changes, marking this conversation as resolved.

zax-ftw commented on include/RE/E/ExtraMapMarker.h:
```diff
@@ -8,14 +8,14 @@ namespace RE
 	enum class MARKER_TYPE
 	{
-		kNone = 0,
+		kEmpty = 0,
```
Copied from Flash with other decompiled names. Looked like the most reliable data we have.

zax-ftw replied:
@Ryan-rsm-McKenzie hmm actually this one is debatable, CK has it defined as `None` and flash as `Empty`, the other markers from this PR are accurate with flash and CK. Should I revert this one to `None` before I push fixed version?

zax-ftw commented on include/RE/E/ExtraMapMarker.h:
```diff
@@ -61,12 +61,21 @@ namespace RE
 		kFalkreathCapitol = 50,
 		kDawnstarCastle = 51,
 		kDawnstarCapitol = 52,
-		kDLC02_TempleOfMiraak = 53,
-		kDLC02_RavenRock = 54,
-		kDLC02_BeastStone = 55,
-		kDLC02_TelMithryn = 56,
-		kDLC02_ToSkyrim = 57,
-		kDLC02_ToSolstheim = 58
```
Copied from `map.swf` with other decompiled names, looked pretty reliable to me.

zax-ftw commented on include/RE/T/TESQuest.h:
```diff
@@ -85,27 +101,11 @@ namespace RE
 	struct QUEST_DATA  // DNAM
 	{
 	public:
-		enum class Type
-		{
-			kNone = 0,
-			kMainQuest = 1,
-			kMagesGuild = 2,
-			kThievesGuild = 3,
-			kDarkBrotherhood = 4,
-			kCompanionsQuest = 5,
-			kMiscellaneous = 6,
-			kDaedric = 7,
-			kSideQuest = 8,
-			kCivilWar = 9,
-			kDLC01_Vampire = 10,
-			kDLC02_Dragonborn = 11
-		};
-
```
Consistency with `QuestFlag` which is also placed outside of class (also similar naming of enum + short variable name in `QUEST_DATA` struct). Its also easily accessible as return type for the purpose of `QuestType GetType` function.

zax-ftw commented on src/RE/T/TESQuest.cpp:
```diff
@@ -22,6 +29,16 @@ namespace RE
 		return currentStage;
 	}
 
+	std::int8_t TESQuest::GetPriority() const
+	{
+		return data.priority;
+	}
+
+	QuestType TESQuest::GetType() const
+	{
+		return data.type.get();
+	}
+
```
Added them in source file because other similarly short functions were also defined here.

Ryan-rsm-McKenzie replied:
Please move them to the header anyway, these files are old and they do not reflect best practices for compile time health in a project as large as this.

zax-ftw commented:
Added all requested changes, reverted `kEmpty` to `kNone` in `MapMarkerData` enum. Added `IsNamed` function to `TESQuest` class (hope you don't mind). Also fixed the class definition of `MapMarkerData` to inherit from `TESFullName` because this is most likely the correct structure: ctor at `001433D0` its even more obvious in Skyrim32 ASM.

zax-ftw commented:
Added important hotfix to `MapMarkerData` with swapped `type` and `pad12` positions (they were declared in wrong order originally, pre this PR).

Ryan-rsm-McKenzie commented on include/RE/E/ExtraMapMarker.h:
```diff
@@ -61,15 +61,24 @@ namespace RE
 	};
 
-	class MapMarkerData
+	class MapMarkerData : public TESFullName
```
`MapMarkerData` does not inherit from `TESFullName`

zax-ftw replied:
I called `TESFullName` functions on it and it worked fine. It looks nicer and is less clunky to use: `marker->GetFullName` vs `marker->locationName.GetFullName()` also I've never seen `TESFullName` being used as standalone member like that?

Ryan-rsm-McKenzie replied:
`MapMarkerData` would have a vtable if it inherited from `TESFullName`. This library is primarily about correctness, please revert this change.

zax-ftw replied:
This change is reverted, marking this conversation as resolved.

zax-ftw commented on include/RE/T/TESQuest.h:
```diff
@@ -41,6 +41,22 @@ namespace RE
 		kHasDialogueData = 1 << 15
 	};
 
+	enum class QuestType
```
Noticed you name some enums with uppercase and some with pascal case ( [Google style](https://google.github.io/styleguide/cppguide.html#Enumerator_Names) ). Which convention should I follow in future PRs?

Ryan-rsm-McKenzie replied:
It depends on the enum

zax-ftw replied:
Should we also change `QuestFlag` to upper for consistency? I was also thinking if `QUEST_OBJECTIVE_STATE` couldn't be brought to `TESQuest` for easiness of use (these three are strictly related) and kept in same naming style: `QuestFlag` `QuestType` `ObjectiveState` or all upper if you prefer?

Ryan-rsm-McKenzie replied:
Please make this change

zax-ftw commented on include/RE/E/ExtraMapMarker.h:
```diff
@@ -61,12 +61,21 @@ namespace RE
+		kDLC02CastleKarstaag = 59,
+		kTotalLocationTypes,
+
+		kDoor,
```
They have it defined like that in flash: `[ ... "DLC02ToSolstheimMarker", "DLC02CastleKarstaagMarker", "", "DoorMarker", "QuestTargetMarker", "QuestTargetDoorMarker", "MultipleQuestTargetMarker" ... ]` - notice the empty space between castle and door marker.

zax-ftw commented:
Reverted `TESFullName` change, reordered `FileExists` definition. I think that my suggestion about quest related enums is worth considering, waiting for your answer before pushing changes.

zax-ftw commented:
Added `GetName` and changed some function defs in `MapMarkerData` because they were extremely cluttered (`constexpr` didn't do anything here and `noexcept` was overkill).

zax-ftw commented:
Since you didn't answer my question, here's my proposed changes to `ObjectiveState` and objective related enums. In my opinion its more readable when its all in same place and using same naming scheme, but let me know if I should revert this commit.

Ryan-rsm-McKenzie commented on include/RE/Q/QuestObjectiveStates.h:
```diff
@@ -1,14 +0,0 @@
-
-namespace RE
-{
-	enum class QUEST_OBJECTIVE_STATE
-	{
-		kDormant = 0,
-		kDisplayed = 1,
-		kCompleted = 2,
-		kCompletedDisplayed = 3,
-		kFailed = 4,
-		kFailedDisplayed = 5
-	};
```
Why did you change this? Please don't change code for no reason.

zax-ftw replied:
@Ryan-rsm-McKenzie It was the only quest related enum that wasn't in TESQuest.h so mostly for consistency and reducing the number of tiny unnecessary files. The uppercase to lower was more of subjective change (PascalCase seems to be more commonly used for enums nowadays) but also for consistency with other quest related enum names.

Ryan-rsm-McKenzie replied:
This is in its own file to prevent circular dependencies. Please revert this change, and in the future do not change files withy a good reason.

zax-ftw replied:
@Ryan-rsm-McKenzie reverted

zax-ftw commented:
FYI: the requested changes were reverted in last commit. It's all fine now?

Ryan-rsm-McKenzie commented on include/RE/E/ExtraMapMarker.h:
```diff
@@ -85,15 +94,16 @@ namespace RE
 
-		constexpr void SetHidden(bool a_value) noexcept { a_value ? flags.set(Flag::kShowAllHidden) : flags.reset(Flag::kShowAllHidden); }
-		constexpr void SetVisible(bool a_value) noexcept { a_value ? flags.set(Flag::kVisible) : flags.reset(Flag::kVisible); }
+		const char* GetName() { return locationName.GetFullName(); }
```
```suggestion
		[[nodiscard]] const char* GetName() { return locationName.GetFullName(); }
```

Ryan-rsm-McKenzie commented on include/RE/T/TESQuest.h:
```diff
@@ -233,11 +233,15 @@ namespace RE
 		TESCondition*                            QConditions() override;                                         // 3D - { return &objConditions; }
 		BGSStoryManagerTreeVisitor::VisitControl AcceptVisitor(BGSStoryManagerTreeVisitor& a_visitor) override;  // 3E
 
+		void          ForceRefIntoAliasID(std::uint32_t a_alias, TESObjectREFR* a_refr);
```
Please order this alphabetically

Ryan-rsm-McKenzie commented on include/RE/T/TESQuest.h:
```diff
@@ -233,11 +233,15 @@ namespace RE
+		void          ForceRefIntoAliasID(std::uint32_t a_alias, TESObjectREFR* a_refr);
 		bool          EnsureQuestStarted(bool& a_result, bool a_startNow);
 		std::uint16_t GetCurrentStageID() const;
+		std::int8_t   GetPriority() const { return data.priority; }
+		QuestType     GetType() const { return data.type.get(); }
 		bool          IsActive() const;
 		bool          IsCompleted() const;
 		bool          IsEnabled() const;
+		bool          IsNamed() const;
```
```suggestion
		[[nodiscard]] std::int8_t   GetPriority() const noexcept { return data.priority; }
		[[nodiscard]] QuestType     GetType() const noexcept { return data.type.get(); }
		bool          IsActive() const;
		bool          IsCompleted() const;
		bool          IsEnabled() const;
		[[nodiscard]] bool          IsNamed() const;
```

Ryan-rsm-McKenzie commented on src/RE/T/TESQuest.cpp:
```diff
@@ -37,6 +44,15 @@ namespace RE
 
+	bool TESQuest::IsNamed() const
+	{
+		auto name = this->GetName();
+		if (name && strlen(name) != 0) {
+			return true;
+		}
+		return false;
```
```suggestion
		const auto name = this->GetName();
		return name && name[0] != '\0';
```

zax-ftw commented:
Wondering why its suddenly failing on build check if I didn't add anything other than suggested changes?
